draft-wang-ppm-dap-taskprov.md

```diff
@@ -222,144 +225,212 @@ struct {
 > discussion.
 
-If aggregator supports `task-prov` extension, it should proceed to
-{{provisioning-a-task}}. If task provision failed, leader MUST alert the client
-with error from {{provisioning-a-task}}. If task provision succeeded, leader
-should continue to the rest of upload flow.
+Otherwise, if the Leader does support the extension, it first attemps to parse
+the payload. If parsing fails, it MUST abort with "unrecognizedMessage".
+
+Next, it checks that the task ID included in the report matches the task ID
+derived from the extension payload as specified in {{construct-task-id}}. If the
+task ID does not match, then the Leader MUST abort with "unrecognizedTask".
 
-Leader MAY return error to the client if task creation failed.
+The Leader then decides whether to opt in to the task as described in
+{{provisioning-a-task}}. If it opts out, it MUST abort the upload request with
+"invalidTask".
+
+> OPEN ISSUE: In case of opt-out, would it be useful to specify how to report
+> this to the Author?
+
+Finally, once the Leader has opted in to the task, it completes the upload
+request as usual.
+
+## Aggregate Protocol
+
+When the Leader opts in to a task, it derives the VDAF verification key for that
+task as described in {{vdaf-verify-key}}.
+
+## Collect Protocol
+
+The Collector might issue a collect request for a task provisioned by the
+`task_prov` extension prior to opting in to the task. In this case, the Leader
+would need to abort the collect request with "unrecognizedTask". When it does
+so, it SHOULD also include a "Retry-After" header in its HTTP response
+indicating the time after which the Collector should retry its request.
+
+> TODO: Find RFC reference for "Retry-After".
+
+> OPEN ISSUE: This semantics is awkward, as there's no way for the Leader to
+> distinguish between Collectors who support the extension and those that don't.
+> We should consider adding an extension field to `CollectReq`.
 
 # Helper Behavior
 
-Helper should have saved any parameters described in {{out-of-band-parameters}}.
-Similar to leader, helper may not know the task ID of the current task before
-receiving the first `AggregateInitializeReq` message from the leader.
+Upon receiving of an `AggregateInitializeReq` from the Leader, If the Helper
+does not support the `task_prov` extension, it MUST ignore the extension
+payload and process each `ReportShare` as usual. In particular, if the Helper
+does not recognize the task ID, it MUST abort the aggregate request with error
+"unrecognizedTask". Otherwise, if the Helper supports the extension, it proceeds
+as follows.
+
+First, the Helper checks that all report shares carried by the request pertain
+to the same task. In particular, it checks that:
+
+1. Either all report shares have the `task_prov` extension or none do. If not
+   the Helper MUST abort with "unrecognizedMessage".
+
+1. All report shares with the `task_prov` extension have the same payload. If
```

Comment:
This should be "2. " right?

Reply:
`make` renders this as an enumerated list (1, 2, 3, ...). Using "1." to indicate an element of an enumerated list is better because you don't have to renumber the list when adding/removing items.